src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+  Useful to fix definitional equalities. -/
+def copy (f : gen_loop N x) (g : cube N → X) (h : g = f) : gen_loop N x :=
+⟨⟨g, h.symm ▸ f.1.2⟩, by { convert f.2, ext1, simp_rw h, refl }⟩
+
+lemma coe_copy (f : gen_loop N x) {g : cube N → X} (h : g = f) : ⇑(copy f g h) = g := rfl
+
+lemma copy_eq (f : gen_loop N x) {g : cube N → X} (h : g = f) : copy f g h = f :=
+by { ext x, exact congr_fun h x }
+
+lemma boundary (f : gen_loop N x) : ∀ y ∈ cube.boundary N, f y = x := f.2
+
+instance fun_like : fun_like (gen_loop N x) (cube N) (λ _, X) :=
 { coe := λ f, f.1,
   coe_injective' := λ ⟨⟨f, _⟩, _⟩ ⟨⟨g, _⟩, _⟩ h, by { congr, exact h } }
 
-@[ext] lemma ext (f g : gen_loop n x) (H : ∀ y, f y = g y) : f = g := fun_like.ext f g H
+@[ext] lemma ext (f g : gen_loop N x) (H : ∀ y, f y = g y) : f = g :=
+fun_like.coe_injective' (funext H)
 
-@[simp] lemma mk_apply (f : C(I^n, X)) (H y) : (⟨f, H⟩ : gen_loop n x) y = f y := rfl
+@[simp] lemma mk_apply (f : C(cube N, X)) (H y) : (⟨f, H⟩ : gen_loop N x) y = f y := rfl
 
-/--
-The constant `gen_loop` at `x`.
--/
-def const : gen_loop n x := ⟨continuous_map.const _ x, λ _ _, rfl⟩
+/-- The constant `gen_loop` at `x`. -/
+def const : gen_loop N x := ⟨continuous_map.const _ x, λ _ _, rfl⟩
 
-instance inhabited : inhabited (gen_loop n x) := { default := const }
+@[simp] lemma const_apply {t} : (@const X _ N x) t = x := rfl
 
-/--
-The "homotopy relative to boundary" relation between `gen_loop`s.
--/
-def homotopic (f g : gen_loop n x) : Prop :=
-f.to_continuous_map.homotopic_rel g.to_continuous_map (cube.boundary n)
+instance inhabited : inhabited (gen_loop N x) := ⟨const⟩
+
+/-- The "homotopy relative to boundary" relation between `gen_loop`s. -/
+def homotopic (f g : gen_loop N x) : Prop := f.1.homotopic_rel g.1 (cube.boundary N)
 
 namespace homotopic
 section
-variables {f g h : gen_loop n x}
+variables {f g h : gen_loop N x}
 
-@[refl] lemma refl (f : gen_loop n x) : homotopic f f := continuous_map.homotopic_rel.refl _
+@[refl] lemma refl (f : gen_loop N x) : homotopic f f := continuous_map.homotopic_rel.refl _
 
-@[symm] lemma symm (H : f.homotopic g) : g.homotopic f := H.symm
+@[symm] lemma symm (H : homotopic f g) : homotopic g f := H.symm
 
-@[trans] lemma trans (H0 : f.homotopic g) (H1 : g.homotopic h) : f.homotopic h := H0.trans H1
+@[trans] lemma trans (H0 : homotopic f g) (H1 : homotopic g h) : homotopic f h := H0.trans H1
 
-lemma equiv : equivalence (@homotopic X _ n x) :=
+lemma equiv : equivalence (@homotopic X _ N x) :=
 ⟨homotopic.refl, λ _ _, homotopic.symm, λ _ _ _, homotopic.trans⟩
 
-instance setoid (n : ℕ) (x : X) : setoid (gen_loop n x) := ⟨homotopic, equiv⟩
+instance setoid (N) (x : X) : setoid (gen_loop N x) := ⟨homotopic, equiv⟩
 
 end
 end homotopic
 
+section
+
+variable [decidable_eq N]
+
+/-- Path from a generalized loop by currying $I^N → X$ into $I → (I^{N\setminus\{j\}} → X)$. -/
+@[simps] def to_path (i : N) : gen_loop N x → Ω (gen_loop {j // j ≠ i} x) const := λ p,
```

Comment:
This should probably be renamed `to_loop` since the codomain is the loop space, and same for `from_path`. Then we should systematically change "path" to "loop" in the docstrings, except when there is no restriction on boundary behavior.

Reply:
Done!

Reply:
Thanks! 

maintainer merge